### ISSUE TITLE
Has the behavior of S3Uploader changed?

### ISSUE THREAD

**Issue by selbyk** [benefitcloud/ember-uploader#55]
I have a working copy behaving the way README.md describes but after recloning my project and npm updating, now it's behaving all wacky and am not sure why.

There is a GET to the signing URL:

```curl
curl 'http://localhost:4200/api/admin/files/sign?name=Brown+Bear+Picture+001.jpg&type=image%2Fjpeg&size=66854'
-H 'Accept-Encoding: gzip, deflate, sdch'
-H 'Accept-Language: en-US,en;q=0.8'
-H 'Authorization: Bearer **************************'
-H 'Content-Type: json'
-H 'X-Requested-With: XMLHttpRequest'
-H 'If-None-Match: W/"2fd-70d78482"'
-H 'Connection: keep-alive'
```

With a proper response:

```json
{
  "acl":"public-read",
  "awsaccesskeyid": "myid",
  ...
  "success_action_status":"201"
}
```

However, now there is no PUT request to the S3 server and the `response`  param in the code block below the is the JSON response from http://localhost:4200/api/admin/files/sign. 

```js
import FileField from 'ember-uploader/file-field';
import S3Uploader from 'ember-uploader/s3';
...
    url: '/api/admin/files/sign',
    filesDidChange: (function () {
        var _this = this;
        var uploadUrl = this.get('url');
        var files = this.get('files');

        var uploader = S3Uploader.create({
          url: uploadUrl
        });

        uploader.on('didUpload', function (response) {
          console.log(JSON.stringify(response)); // JSON from http://localhost:4200/api/admin/files/sign
          // S3 should return XML with url
          var uploadedUrl = $(response).find('Location')[0].textContent; // this throws Uncaught TypeError: Cannot read property 'textContent' of undefined 
          ...
        });
    }.observes('files'),
});
```

Any ideas?

So far I've tried different signing keys and turning off my content security policy, but I'm stumped.

Before I use any outdated package(s), I figured I ask here.

Any help is appreciated!

**Comment by selbyk** [comment]
Node/Express backend implementation, jic.  I'm pretty sure this is working, though.

```js
// Import required modules
var crypto = require('crypto'),
      shasum = crypto.createHash('sha1'),
      base64 = function(str){
          var buf = new Buffer(str).toString('base64');
          return buf;
      };

// Setup AWS Variables
var S3_BUCKET_NAME = 'bucket',
      S3_SECRET_KEY = process.env.AWS_SECRET_KEY,
      S3_ACCESS_KEY = process.env.AWS_ACCESS_KEY;

// Export Express.js routes
module.exports = function(app) {
  app.get('/api/admin/files/sign', function (req, res) {
    var params = {
        "name": req.query.name,
        "size": req.query.size,
        "type": req.query.type
    };

    // PUT request to S3 must start within 5 minutes
    var expires = new Date();
    expires.setMinutes(expires.getMinutes() + 5); // 5 minute later

    var objectName = params.name,
        amzHeaders = "x-amz-acl:public-read",
        mimeType = params.type;

    var bucketKey = "uploads/" + params.name; // directory/file path

    var policy= {
      expiration: expires,
      conditions: [
        { bucket: S3_BUCKET_NAME },
        { acl: 'public-read' },
        { expires: expires },
        { success_action_status: '201' },
        [ 'starts-with', '$key', '' ],
        [ 'starts-with', '$Content-Type', '' ],
        [ 'starts-with', '$Cache-Control', '' ],
        [ 'content-length-range', 0, 524288000 ]
      ]
    };

    var policy64 = base64(JSON.stringify(policy));

    var hash, hmac;

    hmac = crypto.createHmac('sha1', S3_SECRET_KEY);

    // change to 'binary' if you want a binary digest
    hmac.setEncoding('base64');

    // write in the text that you want the hmac digest for
    hmac.write(policy64);

    // you can't read from the stream until you call end()
    hmac.end(function(){
      // read out hmac digest
      hash = hmac.read();
      //console.log("The URL is", url);
      console.log('file ' + params.type + ' signed expires ' + expires);
      res.send({
          "acl": "public-read",
          "awsaccesskeyid": S3_ACCESS_KEY,
          "bucket": S3_BUCKET_NAME,
          "region": "us-west-2",
          "Cache-Control": "max-age=259200, public",
          //"AWSAccessKeyId": S3_ACCESS_KEY,
          "Content-Type": params.type,
          "expires": expires.toISOString(),
          "key": bucketKey,
          "policy": policy64,
          "signature": hash,
          "success_action_status": "201"
      });
    });
  });
...
```

**Comment by selbyk** [comment]
`"ember-uploader": "0.3.4"` in bower.json works, ember-uploader 0.3.5 breaks functionality.  Is there a reason for this?  If so, could you update the readme?  Thanks!

**Comment by digitaltoad** [comment]
I will look into this. There was a change to the way the promise was resolved/rejected and I believe this is the culprit.

**Comment by rahult** [comment]
Getting the same issue with 0.3.5

**Comment by selbyk** [comment]
If I had to guess, it looks like this change is the culprit: https://github.com/benefitcloud/ember-uploader/compare/c5735be1263cf07aec374addefa429b34ce61eee...3ffcef179bb84200bf5b9c8530ad2e1e09dded4e#diff-7da8f6c08d157e1189a4f51cec1e5014R33

**Comment by selbyk** [comment]
https://github.com/benefitcloud/ember-uploader/compare/c5735be1263cf07aec374addefa429b34ce61eee...3ffcef179bb84200bf5b9c8530ad2e1e09dded4e#diff-cd19b38f35e5030671d7f73532b875bdR36 *

**Comment by digitaltoad** [comment]
Can you try 2de1ba4

**Comment by rahult** [comment]
Awesome 2de1ba4 worked :+1:

**Comment by selbyk** [comment]
Oops, didn't notice your reply.  Will test it out when I get back from class. ^^

**Comment by digitaltoad** [comment]
If it works for you, I'll go ahead and make a new build.

**Comment by selbyk** [comment]
Works like a charm! Sorry for the delayed response.

I didn't realize you could just bower install by commit.  Pretty nifty.  Thanks for looking into it. ^^

**Comment by luiscoimbra** [comment]
Hey guys, I had the same problem with 0.3.5 and fixed by changing to 0.3.4.
I got another issue now, the S3 server respond with 
<Message>Invalid according to Policy: Policy Condition failed: ["eq", "$butcket", "my_bucket_name"]</Message>
Does anyone knows what can be? (Is there any specific configuration to S3?)

**Comment by digitaltoad** [comment]
Fixed in 2de1ba4ca6fa4f7c2df1707f583d4d1a761aa6b4

**Comment by digitaltoad** [comment]
New version has been released.
